File: DP/Word/wildcard-matching.py
'''

Description
Implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).
The matching should cover the entire input string (not partial).

Have you met this question in a real interview?
Example
isMatch("aa","a") → false
isMatch("aa","aa") → true
isMatch("aaa","aa") → false
isMatch("aa", "*") → true
isMatch("aa", "a*") → true
isMatch("ab", "?*") → true
isMatch("aab", "c*a*b") → false

'''

class Solution:
  """
  @param s: A string
  @param p: A string includes "?" and "*"
  @return: is Match?
  """
  def isMatch(self, s, p):
    # write your code here
    n = len(s)
    m = len(p)
    f = [[False for _ in range(m + 1)] for _ in range(n + 1)]
    f[0][0] = True

    for i in range(0, n + 1):
      for j in range(0, m + 1):
        if i > 0 and j > 0:
          f[i][j] |= f[i - 1][j - 1] and \
                     (s[i - 1] == p[j - 1] or p[j - 1] in ['?', '*'])

        if i > 0 and j > 0:
          f[i][j] |= f[i - 1][j] and p[j - 1] == '*'

        if j > 0:
          f[i][j] |= f[i][j - 1] and p[j - 1] == '*'

    return f[n][m]

File: real-life-technical/Regex-matching/wildcard-matching.ts
/**
 * 44. Wildcard Matching
 * https://leetcode.com/problems/wildcard-matching/
 *
 * Given an input string (s) and a pattern (p),
 *  implement wildcard pattern matching with support for '?' and '*' where:
 *  1. '?' Matches any single character.
 *  2. '*' Matches any sequence of characters (including the empty sequence).
 * The matching should cover the entire input string (not partial).
 *
 * Example 1:
 * Input: s = "aa", p = "a"
 * Output: false
 * Explanation: "a" does not match the entire string "aa".
 *
 * Example 2:
 * Input: s = "aa", p = "*"
 * Output: true
 * Explanation: '*' matches any sequence.
 *
 *
 * Example 3:
 * Input: s = "cb", p = "?a"
 * Output: false
 * Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.
 *
 * Example 4:
 *  Input: s = "adceb", p = "*a*b"
 * Output: true
 * Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".
 *
 * Example 5:
 * Input: s = "acdcb", p = "a*c?b"
 * Output: false
 *
 * Constraints:
 * 0 <= s.length, p.length <= 2000
 * s contains only lowercase English letters.
 * p contains only lowercase English letters, '?' or '*'.
 */

function isMatch(s: string, p: string): boolean {
  // 2D map for memorization.
  const dp = {};

  // Helper function.
  function helper(s: string, p: string): boolean {
    if (dp[s] && dp[s][p] !== undefined) {
      return dp[s][p];
    }

    // Handle duplicate * cases.
    if (p[0] === "*" && p[1] === "*") {
      return helper(s, p.slice(1));
    }

    if (dp[s] === undefined) {
      dp[s] = {};
    }

    if (s === p) {
      dp[s][p] = true;
    } else if (p === "*") {
      dp[s][p] = true;
    } else if (s === "" || p[0] === "") {
      dp[s][p] = false;
    } else if (s[0] === p[0] || p[0] === "?") {
      dp[s][p] = helper(s.slice(1), p.slice(1));
    } else if (p[0] === "*") {
      dp[s][p] = helper(s.slice(1), p) || helper(s, p.slice(1));
    } else {
      dp[s][p] = false;
    }

    return dp[s][p];
  }
  // console.log(dp);
  return helper(s, p);
}

// console.log(isMatch('aa', 'a'));
// console.log(isMatch('aa', '*'));
// console.log(isMatch('cb', '?a'));
// console.log(isMatch('adceb', '*a*b'));
// console.log(isMatch('acdcb', 'a*c?b'));
console.log(isMatch("", "******"));
